src/parser.rs

```diff
@@ -112,145 +112,216 @@ impl Library {
-
-                    fields.push(Field {
-                        name: field_name,
-                        typ: Type::union(self, u, ns_id),
-                        doc: u_doc,
-                        c_type: ctype,
-                        ..Field::default()
-                    });
-                    union_count += 1;
-                }),
-            "attribute" => parser.ignore_element(),
-            _ => Err(parser.unexpected_element(elem)),
-        })?;
+            let f = self.read_function(ns_id, constructor, FunctionKind::Constructor)?;
+            fns.push(f);
+        }
+        for function in elem.functions() {
+            if function.moved_to().is_some() {
+                continue;
+            }
+            let f = self.read_function(ns_id, function, FunctionKind::Function)?;
+            fns.push(f);
+        }
+        for method in elem.methods() {
+            if method.moved_to().is_some() {
+                continue;
+            }
+            let f = self.read_method(ns_id, method)?;
+            fns.push(f);
+        }
+        for signal in elem.signals() {
+            let signal = self.read_signal(ns_id, signal)?;
+            signals.push(signal);
+        }
+        for property in elem.properties() {
+            properties.push(self.read_property(ns_id, property, &symbol_prefix)?);
+        }
+
+        for implement in elem.implements() {
+            impls.push(self.find_or_stub_type(ns_id, implement.name()));
+        }
+
+        for field in elem.fields() {
+            fields.push(self.read_field(ns_id, field)?);
+        }
+
+        for virtual_method in elem.virtual_methods() {
+            if virtual_method.moved_to().is_some() {
+                continue;
+            }
+            let f = self.read_virtual_method(ns_id, virtual_method)?;
+            vfns.push(f);
+        }
+
+        for union in elem.unions() {
+            let mut u = self.read_union(ns_id, union, Some(&name), Some(&c_type))?;
+            let field_name = if let Some(field_name) = union.name() {
+                field_name.into()
+            } else {
+                format!("u{union_count}")
+            };
+
+            u = Union {
+                name: format!("{name}_{field_name}"),
+                c_type: Some(format!("{c_type}_{field_name}")),
+                ..u
+            };
+
+            let u_doc = u.doc.clone();
+            let ctype = u.c_type.clone();
+
+            fields.push(Field {
+                name: field_name,
+                typ: Type::union(self, u, ns_id),
+                doc: u_doc,
+                c_type: ctype,
+                ..Field::default()
+            });
+            union_count += 1;
+        }
 
-        let parent = elem
-            .attr("parent")
-            .map(|s| self.find_or_stub_type(ns_id, s));
+        let parent = elem.parent().map(|s| self.find_or_stub_type(ns_id, s));
         let typ = Type::Class(Class {
```

Comment:
Instead of using custom types for these things, could we move over to just using directly the types provided by your parser?

Reply:
That would require a huge amount of changes to be done before doing the switch. I rather do that as a second step and my intention for now is to make the gir types wrappers of the ones from the types provided by the parser + TypeId + fields for overrides the type sets

Reply:
That makes sense